vaadin-spring/src/main/java/com/vaadin/flow/spring/scopes/VaadinRouteScope.java

```diff
@@ -0,0 +1,380 @@
+/*
+ * Copyright 2000-2021 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.spring.scopes;
+
+import javax.servlet.ServletContext;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.springframework.beans.factory.ObjectFactory;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.context.ApplicationContext;
+import org.springframework.web.context.support.WebApplicationContextUtils;
+
+import com.vaadin.flow.component.ComponentEventListener;
+import com.vaadin.flow.component.ComponentUtil;
+import com.vaadin.flow.component.DetachEvent;
+import com.vaadin.flow.component.HasElement;
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.component.page.ExtendedClientDetails;
+import com.vaadin.flow.router.AfterNavigationEvent;
+import com.vaadin.flow.router.AfterNavigationListener;
+import com.vaadin.flow.router.BeforeEnterEvent;
+import com.vaadin.flow.router.BeforeEnterListener;
+import com.vaadin.flow.router.RouterLayout;
+import com.vaadin.flow.server.UIInitEvent;
+import com.vaadin.flow.server.UIInitListener;
+import com.vaadin.flow.server.VaadinContext;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.VaadinServletContext;
+import com.vaadin.flow.server.VaadinSession;
+import com.vaadin.flow.shared.Registration;
+import com.vaadin.flow.spring.SpringVaadinSession;
+import com.vaadin.flow.spring.annotation.RouteScopeOwner;
+
+/**
+ * Implementation of Spring's
+ * {@link org.springframework.beans.factory.config.Scope} that binds the beans
+ * to a component in the navigation chain. Registered by default as the scope "
+ * {@value #VAADIN_ROUTE_SCOPE_NAME}".
+ *
+ * @see com.vaadin.flow.spring.annotation.VaadinSessionScope
+ * @author Vaadin Ltd
+ * @since
+ *
+ */
+public class VaadinRouteScope extends AbstractScope implements UIInitListener {
+
+    public static final String VAADIN_ROUTE_SCOPE_NAME = "vaadin-route";
+
+    private static class RouteStoreWrapper implements Serializable {
```

Comment:
Non-blocking: I think `SessionRouteStore` or such would have been a more descriptive name than just _wrapper_

Reply:
Every "store" is stored inside a session. So `SessionRouteStore` name doesn't bring a new information.

`RouteStoreWrapper` is not a best name but I can't suggest any better and I don't see `SessionRouteStore`  as a better name.

Reply:
I disagree, it would add more value as routes are generally registered to application scope and secondarily to session scope, even while here for routescope implementation, it only makes sense to store beans to session. I dislike "wrapper" naming because it there is typically reason why things are wrapped, and here is not even a wrapper but a holder of the route stores per specific session. Thus it is like `RouteStoreStore` / or `SessionRouteStoreMap` orwhatnot. But like I said, it was not blocking.